File: HR APP BACKEND/app/schemas/a2a.py
from __future__ import annotations

from datetime import datetime, timezone
from enum import Enum
from typing import Any, Literal
from uuid import uuid4

from pydantic import BaseModel, ConfigDict, Field, model_validator


def utcnow() -> datetime:
    return datetime.now(timezone.utc)


class A2ABaseModel(BaseModel):
    model_config = ConfigDict(extra="forbid")


class TaskStatus(str, Enum):
    queued = "queued"
    running = "running"
    completed = "completed"
    failed = "failed"
    canceled = "canceled"


class TaskExecutionMode(str, Enum):
    sync = "sync"
    async_ = "async"


class Capability(A2ABaseModel):
    name: str = Field(min_length=2, max_length=80)
    description: str = Field(min_length=2, max_length=500)
    input_modes: list[str] = Field(default_factory=lambda: ["text", "json"], max_length=10)
    output_modes: list[str] = Field(default_factory=lambda: ["json"], max_length=10)
    streaming: bool = False
    side_effects: bool = False


class Skill(A2ABaseModel):
    id: str = Field(min_length=2, max_length=80, pattern=r"^[a-zA-Z0-9_.:-]+$")
    name: str = Field(min_length=2, max_length=120)
    description: str = Field(min_length=2, max_length=800)
    tags: list[str] = Field(default_factory=list, max_length=20)
    examples: list[str] = Field(default_factory=list, max_length=10)


class AgentCard(A2ABaseModel):
    id: str = Field(min_length=2, max_length=80, pattern=r"^[a-zA-Z0-9_.:-]+$")
    name: str = Field(min_length=2, max_length=120)
    description: str = Field(min_length=2, max_length=1200)
    protocol_version: str = "a2a-0.1"
    version: str = "1.0.0"
    url: str = Field(min_length=1, max_length=500)
    provider: str = "HIREAI"
    visibility: Literal["public", "hr", "internal"] = "hr"
    enabled: bool = True
    capabilities: list[Capability] = Field(default_factory=list, max_length=20)
    skills: list[Skill] = Field(default_factory=list, max_length=20)
    default_input_modes: list[str] = Field(default_factory=lambda: ["text", "json"])
    default_output_modes: list[str] = Field(default_factory=lambda: ["json"])
    auth_schemes: list[str] = Field(default_factory=lambda: ["bearer"])
    metadata: dict[str, Any] = Field(default_factory=dict)


class TraceMetadata(A2ABaseModel):
    trace_id: str = Field(default_factory=lambda: str(uuid4()), min_length=8, max_length=128)
    request_id: str | None = Field(default=None, max_length=128)
    run_id: str | None = Field(default=None, max_length=128)
    correlation_id: str | None = Field(default=None, max_length=128)
    agent_trace: list[dict[str, Any]] = Field(default_factory=list)


class ExecutionMetadata(A2ABaseModel):
    agent_id: str = Field(min_length=2, max_length=80)
    runtime: Literal["hr_multi_agent_runtime"] = "hr_multi_agent_runtime"
    started_at: datetime | None = None
    completed_at: datetime | None = None
    latency_ms: float | None = Field(default=None, ge=0)
    model_used: str | None = Field(default=None, max_length=160)
    token_usage: dict[str, Any] = Field(default_factory=dict)
    status_code: str | None = Field(default=None, max_length=80)


class AgentMessage(A2ABaseModel):
    role: Literal["user", "agent", "system"] = "user"
    content: str = Field(default="", max_length=20000)
    context: dict[str, Any] = Field(default_factory=dict)
    metadata: dict[str, Any] = Field(default_factory=dict)
    task_id: str | None = Field(default=None, max_length=80)
    trace_id: str | None = Field(default=None, max_length=128)

    @model_validator(mode="after")
    def content_or_context_required(self) -> "AgentMessage":
        if not (self.content or "").strip() and not self.context:
            raise ValueError("Either content or context is required")
        return self


class AgentArtifact(A2ABaseModel):
    id: str = Field(default_factory=lambda: str(uuid4()))
    task_id: str = Field(min_length=8, max_length=80)
    name: str = Field(min_length=2, max_length=120)
    artifact_type: Literal["json", "text", "trace", "error"] = "json"
    mime_type: str = Field(default="application/json", max_length=120)
    data: dict[str, Any] | list[Any] | str
    redacted: bool = False
    created_at: datetime = Field(default_factory=utcnow)
    metadata: dict[str, Any] = Field(default_factory=dict)


class TaskResult(A2ABaseModel):
    summary: str = Field(default="", max_length=2000)
    output: dict[str, Any] = Field(default_factory=dict)
    artifact_ids: list[str] = Field(default_factory=list)


class AgentTask(A2ABaseModel):
    id: str = Field(default_factory=lambda: str(uuid4()))
    agent_id: str = Field(min_length=2, max_length=80)
    status: TaskStatus = TaskStatus.queued
    owner_id: str = Field(min_length=1, max_length=120)
    message: AgentMessage
    result: TaskResult | None = None
    error: str | None = Field(default=None, max_length=2000)
    trace: TraceMetadata = Field(default_factory=TraceMetadata)
    execution: ExecutionMetadata
    artifact_ids: list[str] = Field(default_factory=list)
    created_at: datetime = Field(default_factory=utcnow)
    updated_at: datetime = Field(default_factory=utcnow)


class CreateTaskRequest(A2ABaseModel):
    agent_id: str = Field(min_length=2, max_length=80)
    message: AgentMessage
    execution_mode: TaskExecutionMode = TaskExecutionMode.async_


class EvaluationRunRequest(A2ABaseModel):
    resume_text: str = Field(min_length=20, max_length=30000)
    jd_text: str = Field(min_length=20, max_length=30000)
    label: str = Field(default="Resume screening evaluation", max_length=120)
    execution_mode: TaskExecutionMode = TaskExecutionMode.async_


class TaskStatusResponse(A2ABaseModel):
    task_id: str
    agent_id: str
    status: TaskStatus
    error: str | None = None
    updated_at: datetime


class AgentDirectoryResponse(A2ABaseModel):
    agents: list[AgentCard]
    internal_count: int = 0


class A2AAuditEvent(A2ABaseModel):
    id: str = Field(default_factory=lambda: str(uuid4()))
    actor_id: str = Field(min_length=1, max_length=160)
    actor_type: Literal["technical_admin", "service_token"]
    action: str = Field(min_length=2, max_length=120)
    resource: str = Field(min_length=2, max_length=120)
    resource_id: str | None = Field(default=None, max_length=120)
    detail: dict[str, Any] = Field(default_factory=dict)
    created_at: datetime = Field(default_factory=utcnow)


class A2AAuditResponse(A2ABaseModel):
    events: list[A2AAuditEvent]

File: HR APP FRONTEND/src/services/a2a.ts
import api from "@/services/api";

export type A2ATaskStatus = "queued" | "running" | "completed" | "failed" | "canceled";

export interface A2ACapability {
  name: string;
  description: string;
  input_modes: string[];
  output_modes: string[];
  streaming: boolean;
  side_effects: boolean;
}

export interface A2ASkill {
  id: string;
  name: string;
  description: string;
  tags: string[];
  examples: string[];
}

export interface A2AAgentCard {
  id: string;
  name: string;
  description: string;
  protocol_version: string;
  version: string;
  url: string;
  provider: string;
  visibility: "public" | "hr" | "internal";
  enabled: boolean;
  capabilities: A2ACapability[];
  skills: A2ASkill[];
  default_input_modes: string[];
  default_output_modes: string[];
  auth_schemes: string[];
  metadata: Record<string, unknown>;
}

export interface A2AAgentMessage {
  role: "user" | "agent" | "system";
  content: string;
  context: Record<string, unknown>;
  metadata?: Record<string, unknown>;
  task_id?: string | null;
  trace_id?: string | null;
}

export interface A2AExecutionMetadata {
  agent_id: string;
  runtime: "hr_multi_agent_runtime";
  started_at: string | null;
  completed_at: string | null;
  latency_ms: number | null;
  model_used: string | null;
  token_usage: Record<string, unknown>;
  status_code: string | null;
}

export interface A2ATraceMetadata {
  trace_id: string;
  request_id: string | null;
  run_id: string | null;
  correlation_id: string | null;
  agent_trace: Array<Record<string, unknown>>;
}

export interface A2AArtifact {
  id: string;
  task_id: string;
  name: string;
  artifact_type: "json" | "text" | "trace" | "error";
  mime_type: string;
  data: unknown;
  redacted: boolean;
  created_at: string;
  metadata: Record<string, unknown>;
}

export interface A2AAuditEvent {
  id: string;
  actor_id: string;
  actor_type: "technical_admin" | "service_token";
  action: string;
  resource: string;
  resource_id: string | null;
  detail: Record<string, unknown>;
  created_at: string;
}

export interface A2ATask {
  id: string;
  agent_id: string;
  status: A2ATaskStatus;
  owner_id: string;
  message: A2AAgentMessage;
  result: {
    summary: string;
    output: Record<string, unknown>;
    artifact_ids: string[];
  } | null;
  error: string | null;
  trace: A2ATraceMetadata;
  execution: A2AExecutionMetadata;
  artifact_ids: string[];
  created_at: string;
  updated_at: string;
}

export interface ADKShadowSummary {
  enabled: boolean;
  execution_mode: string;
  events: number;
  completed: number;
  failed: number;
  compared: number;
  matched: number;
  match_rate_pct: number | null;
}

export interface ADKShadowEvent {
  workflow: string;
  status: string;
  started_at: string;
  latency_ms: number;
  production_hash: string | null;
  shadow_hash: string | null;
  match: boolean | null;
  execution_mode: string;
  entity_id: string | null;
  actor_id: string | null;
  error: string | null;
  metadata: Record<string, unknown>;
}

export interface ADKPromotionStatus {
  enabled: boolean;
  allowlist: string[];
  effective_workflows: Record<string, boolean>;
  timeout_seconds: number;
  fallback_to_legacy: boolean;
  min_quiz_quality_score: number;
  recent: ADKShadowEvent[];
  recent_counts: {
    completed: number;
    fallback: number;
    failed: number;
  };
}

export async function getA2AAgents(includeInternal = false): Promise<A2AAgentCard[]> {
  const res = await api.get("/a2a/agents", { params: { include_internal: includeInternal } });
  return res.data.agents || [];
}

export async function getA2AAgentCard(agentId: string): Promise<A2AAgentCard> {
  const res = await api.get(`/a2a/agents/${encodeURIComponent(agentId)}/card`);
  return res.data;
}

export async function sendA2AMessage(agentId: string, message: A2AAgentMessage): Promise<A2ATask> {
  const res = await api.post(`/a2a/agents/${encodeURIComponent(agentId)}/message`, message);
  return res.data;
}

export async function createA2ATask(
  agentId: string,
  message: A2AAgentMessage,
  executionMode: "sync" | "async" = "async",
): Promise<A2ATask> {
  const res = await api.post("/a2a/tasks", {
    agent_id: agentId,
    message,
    execution_mode: executionMode,
  });
  return res.data;
}

export async function getA2ATask(taskId: string): Promise<A2ATask> {
  const res = await api.get(`/a2a/tasks/${encodeURIComponent(taskId)}`);
  return res.data;
}

export async function getA2ATaskArtifacts(taskId: string): Promise<A2AArtifact[]> {
  const res = await api.get(`/a2a/tasks/${encodeURIComponent(taskId)}/artifacts`);
  return res.data || [];
}

export async function downloadA2AArtifact(taskId: string, artifactId: string): Promise<void> {
  const res = await api.get(
    `/a2a/tasks/${encodeURIComponent(taskId)}/artifacts/${encodeURIComponent(artifactId)}/download`,
    { responseType: "blob" },
  );
  const disposition = String(res.headers["content-disposition"] || "");
  const match = disposition.match(/filename="([^"]+)"/);
  const filename = match?.[1] || `${artifactId}.json`;
  const url = URL.createObjectURL(res.data);
  const link = document.createElement("a");
  link.href = url;
  link.download = filename;
  document.body.appendChild(link);
  link.click();
  link.remove();
  URL.revokeObjectURL(url);
}

export async function runA2AResumeScreeningEvaluation(payload: {
  resume_text: string;
  jd_text: string;
  label?: string;
  execution_mode?: "sync" | "async";
}): Promise<A2ATask> {
  const res = await api.post("/a2a/evaluations/resume-screening", {
    label: "Resume screening evaluation",
    execution_mode: "async",
    ...payload,
  });
  return res.data;
}

export async function getA2AAudit(limit = 100): Promise<A2AAuditEvent[]> {
  const res = await api.get("/a2a/audit", { params: { limit } });
  return res.data.events || [];
}

export async function getADKShadowSummary(): Promise<ADKShadowSummary> {
  const res = await api.get("/a2a/adk-shadow/summary");
  return res.data;
}

export async function getADKShadowRecent(limit = 50): Promise<ADKShadowEvent[]> {
  const res = await api.get("/a2a/adk-shadow/recent", { params: { limit } });
  return res.data.events || [];
}

export async function getADKPromotionStatus(limit = 25): Promise<ADKPromotionStatus> {
  const res = await api.get("/a2a/adk-promotion/status", { params: { limit } });
  return res.data;
}
